### ISSUE TITLE
Converting Derived Entity Types By Control Info (`@odata.type`)

### ISSUE THREAD

**Issue by brease-colin** [odata2ts/odata2ts#322]
Some context: https://github.com/odata2ts/odata2ts/discussions/320#discussioncomment-10727721

A limited version of our Edmx file is here: 

```
<?xml version="1.0" encoding="UTF-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
	<edmx:DataServices>
		<Schema Namespace="OData.Domain.Entities" xmlns="http://docs.oasis-open.org/odata/ns/edm">
			<EntityType Name="Agenda">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="IsExternal" Type="Edm.Boolean" Nullable="false"/>
				<Property Name="EventOccurrences" Type="Collection(OData.Domain.ComplexTypes.AgendaEventOccurrence)"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Resources" Type="Collection(OData.Domain.Entities.AgendaResourceLink)"/>
			</EntityType>
			<EntityType Name="AgendaResourceLink">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="AgendaId" Type="Edm.Guid"/>
				<Property Name="ResourceId" Type="Edm.Guid" Nullable="false"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Agenda" Type="OData.Domain.Entities.Agenda">
					<ReferentialConstraint Property="AgendaId" ReferencedProperty="Id"/>
				</NavigationProperty>
			</EntityType>
			<EntityType Name="Event" Abstract="true">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="EventType" Type="OData.Domain.Enums.EventType" Nullable="false"/>
				<Property Name="Subject" Type="Edm.String"/>
				<Property Name="Message" Type="Edm.String"/>
				<Property Name="AgendaOccurrences" Type="Collection(OData.Domain.ComplexTypes.AgendaEventOccurrence)" Nullable="false"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Occurrences" Type="Collection(OData.Domain.Entities.EventOccurrence)"/>
			</EntityType>
			<EntityType Name="EventOccurrence">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="Index" Type="Edm.Int32" Nullable="false"/>
				<Property Name="Time" Type="OData.Shared.ComplexTypes.Period" Nullable="false"/>
				<Property Name="EventId" Type="Edm.Guid"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Event" Type="OData.Domain.Entities.Event">
					<ReferentialConstraint Property="EventId" ReferencedProperty="Id"/>
				</NavigationProperty>
			</EntityType>
			<EntityType Name="Absence" BaseType="OData.Domain.Entities.Event"/>
			<EntityType Name="Appointment" BaseType="OData.Domain.Entities.Event">
				<Property Name="LocationId" Type="Edm.Guid"/>
				<NavigationProperty Name="Location" Type="OData.Domain.Entities.Location">
					<ReferentialConstraint Property="LocationId" ReferencedProperty="Id"/>
				</NavigationProperty>
				<NavigationProperty Name="Activities" Type="Collection(OData.Domain.Entities.AppointmentActivity)"/>
			</EntityType>
			<EntityType Name="Location">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="Name" Type="Edm.String" Nullable="false"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
			</EntityType>
			<EntityType Name="AppointmentActivity">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="AppointmentId" Type="Edm.Guid"/>
				<Property Name="ActivityId" Type="Edm.Guid"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Appointment" Type="OData.Domain.Entities.Appointment">
					<ReferentialConstraint Property="AppointmentId" ReferencedProperty="Id"/>
				</NavigationProperty>
				<NavigationProperty Name="Activity" Type="OData.Domain.Entities.Activity">
					<ReferentialConstraint Property="ActivityId" ReferencedProperty="Id"/>
				</NavigationProperty>
			</EntityType>
			<EntityType Name="Activity">
				<Key>
					<PropertyRef Name="Id"/>
				</Key>
				<Property Name="Name" Type="Edm.String" Nullable="false"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
			</EntityType>
			<EntityType Name="AppointmentScheduleItem" BaseType="OData.Domain.Entities.Event">
				<Property Name="LocationId" Type="Edm.Guid"/>
				<NavigationProperty Name="Location" Type="OData.Domain.Entities.Location">
					<ReferentialConstraint Property="LocationId" ReferencedProperty="Id"/>
				</NavigationProperty>
			</EntityType>
			<EntityType Name="Presence" BaseType="OData.Domain.Entities.Event">
				<Property Name="LocationId" Type="Edm.Guid"/>
				<NavigationProperty Name="Location" Type="OData.Domain.Entities.Location">
					<ReferentialConstraint Property="LocationId" ReferencedProperty="Id"/>
				</NavigationProperty>
			</EntityType>
		</Schema>
		<Schema Namespace="OData.Domain.ComplexTypes" xmlns="http://docs.oasis-open.org/odata/ns/edm">
			<ComplexType Name="AgendaEventOccurrence">
				<Property Name="EventId" Type="Edm.Guid" Nullable="false"/>
				<Property Name="AgendaId" Type="Edm.Guid" Nullable="false"/>
				<Property Name="Index" Type="Edm.Int32" Nullable="false"/>
				<Property Name="EventType" Type="OData.Domain.Enums.EventType" Nullable="false"/>
				<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
				<NavigationProperty Name="Event" Type="OData.Domain.Entities.Event"/>
				<NavigationProperty Name="Agenda" Type="OData.Domain.Entities.Agenda"/>
			</ComplexType>
		</Schema>
		<Schema Namespace="OData.Shared.ComplexTypes" xmlns="http://docs.oasis-open.org/odata/ns/edm">
			<ComplexType Name="Period">
				<Property Name="From" Type="Edm.DateTimeOffset" Nullable="false"/>
				<Property Name="To" Type="Edm.DateTimeOffset" Nullable="false"/>
			</ComplexType>
		</Schema>
		<Schema Namespace="OData.Domain.Enums" xmlns="http://docs.oasis-open.org/odata/ns/edm">
			<EnumType Name="EventType">
				<Member Name="Appointment" Value="0"/>
				<Member Name="Absence" Value="1"/>
				<Member Name="Presence" Value="2"/>
				<Member Name="AppointmentScheduleItem" Value="3"/>
			</EnumType>
		</Schema>
		<Schema Namespace="Default" xmlns="http://docs.oasis-open.org/odata/ns/edm">
			<EntityContainer Name="Container">
				<EntitySet Name="Agendas" EntityType="OData.Domain.Entities.Agenda">
					<NavigationPropertyBinding Path="EventOccurrences/Agenda" Target="Agendas"/>
				</EntitySet>
			</EntityContainer>
		</Schema>
	</edmx:DataServices>
</edmx:Edmx>
```

The errors occur on URL's like the following:
`/Agendas(00000000-b439-49bc-a72d-9ab78a08a7e4)/EventOccurrences?$expand=Event($expand=Brease.OData.Domain.Entities.Appointment/Activities)`

Below is the expanded 'event' part of the data that is the result of the query (might not fully match the above schema where I cut some things out, but it paints the picture):

```
{
  "@odata.type": "#OData.Domain.Entities.Appointment",
  "eventType": "Appointment",
  "subject": "Test",
  "message": "",
  "LocationId": null,
  "agendaOccurrences": [
    {
      "eventId": "00000000-818b-4ed5-b28b-353dde3c2f15",
      "etc": "00000000-b439-49bc-a72d-9ab78a08a7e4",
    }
  ],
  "Activities": [
    {
      "AppointmentId": "00000000-818b-4ed5-b28b-353dde3c2f15",
      "ActivityId": "00000000-bc54-4c57-bfca-83c0899f28a4",
      "Id": "00000000-b19e-42bd-9ff7-023aa930adbe"
    }
  ]
}
```

As you can see, all properties that are part of the abstract (Event) class have the correct casing (camel case), but every expanded class (Appointment)'s own properties are still in PascalCase. The generated Typescript interfaces are fine, it's the resulting data where the error is, which is inconsistent with the interface.

My config:
```
mode: Modes.all,
v4BigNumberAsString: true,
allowRenaming: true,
emitMode: EmitModes.ts,
converters: [
    {
        module: '@odata2ts/converter-common',
        use: [ 'dateTimeOffsetToDateConverter', 'simpleDurationConverter'],
    },
],
```

**Comment by texttechne** [comment]
Hi @brease-colin,

that happens because of the casting. Since odata2ts doesn't know, that you are casting the `Event` to an `Appointment`, it uses the Q-object for the event instead of the Q-object for the appointment. And the former doesn't know these properties, hence they are not converted in any way, but just passed through as is (OData requires this behaviour for unknown props).

You can now make the point: Hey, but `allowRenaming` is activated, so odata2ts should also apply the NamingStrategy to unknown props and thereby convert the unknown props to camelCase. However, this wouldn't solve the problem, that an unknown prop cannot be converted regarding its type. So if the unknown prop is an `Edm.DateTimeOffset` it wouldn't get converted to a `Date`. This cannot be fixed as the vital type information is missing.

I would rather propose to rename the ticket: You actually want support for casting.

**Comment by brease-colin** [comment]
Do you mean the casting in the URL? Because if I use the following URL, without the cast in the URL, the same happens, but obviously only for the LocationId as the Activities are not there.

/Agendas(00000000-b439-49bc-a72d-9ab78a08a7e4)/EventOccurrences?$expand=Event

Result:
```
{
  "@odata.type": "#OData.Domain.Entities.Appointment",
  "eventType": "Appointment",
  "subject": "Test",
  "message": "",
  "id": "00000000-818b-4ed5-b28b-353dde3c2f15",
  "LocationId": null
}
```

I just used the casting URL because that was a use case I was in the middle of, but the URL casting part is not part of the problem.

**Comment by texttechne** [comment]
This is essentially the same problem: odata2ts only knows that the response object is of type `Event`, which doesn't contain the `LocationId`. Only the server (and the user) knows what kind of response object is used when an abstract response object is advertised in EDMX.

In your original problem you are forced to use casting in order to realize the expand. However, without casting you still get an object of type `Appointment` while odata2ts assumes `Event`.

So explicit casting by the user would be the proper solution. This way odata2ts would know what to do.

**Comment by texttechne** [comment]
> The user doesn't though as it's a list of items which can all be of different sub types. [...] The EDMX does know it, because it's an abstract class and the Appointment, Absence, Presence classes all have base type Event.

Sorry, I have to admit to be a nitpicker at times 😉: Actually EDMX and the user know the same stuff: It can be one of those types extending from the abstract class. There should be use cases where the user knows more though, because of context. I think your original problem is an example of that.

Just handling abstract response types can be done as you propose: Checking `@odata.type` at runtime and choosing the correct q-object there. Some thoughts:
*  The typing might get interesting though: Should the interface then specify `Array<Event>` or `Array<Appointment|Absence|Presence>`? 
* In the end the casting part is just shifted to the user, when she is handling the response. Imagine what you need to do to make TS happy when using this kind of models...

Regarding your original problem, let me ask you this: How does your OData service behave when you do stuff like that `EventOccurrences?$expand=Event($expand=Brease.OData.Domain.Entities.Appointment/Activities)`?

Does it filter the events by type and then apply the expanding? Or do you also get all other types as well and the property which should be expanded is simply non-existent?

**Comment by brease-colin** [comment]
Thanks for your response!

> The typing might get interesting though: Should the interface then specify `Array<Event>` or `Array<Appointment|Absence|Presence>`?

To me, this makes no difference. I'm used to seeing `Array<Event>` when there's an abstract class (or base interface), but that may also be because I work with languages that do not have the other option. BTW, in my situation it's not an array of events, but an array of event occurrences with each having a single Event link. In that case I would also expect it to be an `Event` property type, not an `Appointment|Absence|Presence` property type. Either way works for me though.

> In the end the casting part is just shifted to the user, when she is handling the response. Imagine what you need to do to make TS happy when using this kind of models...

I'm actually doing that already at the moment and it's what I expected. I use type guards where necessary.

> Regarding your original problem, let me ask you this: How does your OData service behave when you do stuff like that `EventOccurrences?$expand=Event($expand=Brease.OData.Domain.Entities.Appointment/Activities)`?
> 
> Does it filter the events by type and then apply the expanding? Or do you also get all other types as well and the property which should be expanded is simply non-existent?

It behaves just like this example response, which does the latter: https://services.odata.org/V4/(S(xjqbds2oavibr01gt1fny24s))/TripPinServiceRW/People?$expand=Trips($expand=PlanItems($expand=Microsoft.OData.SampleService.Models.TripPin.Flight/Airline))

See `value[0].Trips[0].PlanItems[0]`, which is of type `Flight` and does have the `Airline`, while `value[0].Trips[0].PlanItems[2]` is of type `Event` and has no property `Airline`.

**Comment by texttechne** [comment]
Hi @brease-colin,
this is somewhat hard to get right.

From the [spec](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_ControlInformationtypeodatatype):
> The type control information MUST appear in requests and in responses with [minimal](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_metadataminimalodatametadataminimal) or [full](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_metadatafullodatametadatafull) metadata, if the type cannot be heuristically determined, as described below, and one of the following is true:
·         The type is derived from the type specified for the (collection of) entities or (collection of) complex type instances, or
·         The type is for a property whose type is not declared in $metadata.

This holds true for your use case with abstract response types, of course. Indeed it solves the mystery of abstract response types. But I've got trouble to imagine other use cases based on this definition and the public available odata services. 

To put it differently, my understanding is, that every base type could potentially return a response with `@odata.type` referencing a sub type, right? And most of the time this wouldn't apply, just adding useless code to the generated artefacts...
So for example, given this EDMX:
```xml
      <EntityType Name="Person">
        <!-- ... -->
        <Property Name="HomeAddress" Type="Trippin.Location" />
      </EntityType>
      <ComplexType Name="Location">
        <Property Name="Address" Type="Edm.String" />
        <Property Name="City" Type="Trippin.City" />
      </ComplexType>
      <ComplexType Name="AirportLocation" BaseType="Trippin.Location">
        <Property Name="Loc" Type="Edm.GeographyPoint" />
      </ComplexType>
      <ComplexType Name="EventLocation" BaseType="Trippin.Location">
        <Property Name="BuildingInfo" Type="Edm.String" />
      </ComplexType>
```
From the client side I wouldn't know that the `HomeAddress` always is a `Location` and not a mix of the three ComplexTypes... 

What do you think? Can you do some experiments with your OData service in this regard? 

P.S: Quite interestingly, but also obviously, when using type cast segments, `@odata.type` won't be present: [see this example](https://services.odata.org/V4/(S(xjqbds2oavibr01gt1fny24s))/TripPinServiceRW/People('russellwhyte')/Trips(1003)/PlanItems/Microsoft.OData.SampleService.Models.TripPin.Flight)

**Comment by brease-colin** [comment]
Hi @texttechne ,

Thanks for the extensive reply / research. I've been (and still am) busy for the past week, so I couldn't reply sooner and can't dive in too deeply now either or do experiments. I also see that you have followed up this reply with a pull request, so I look forward to reading what was implemented in this area.

A couple of comments:

> Indeed it solves the mystery of abstract response types. But I've got trouble to imagine other use cases based on this definition and the public available odata services.

I think the main (and only?) use case may be abstract classes indeed, although maybe it doesn't always need to be an abstract class. It can be non-abstract classes as well that are extended? (Yes, you showed that in the next example).

> To put it differently, my understanding is, that every base type could potentially return a response with @odata.type referencing a sub type, right? And most of the time this wouldn't apply, just adding useless code to the generated artefacts...

Only if the type is mentioned at least once as a BaseType, this would be necessary. I wouldn't call that useless code though, but it may feel that way if it only applies to rare cases.

> From the client side I wouldn't know that the HomeAddress always is a Location and not a mix of the three ComplexTypes...

I'm not sure what you mean here. Do you mean to say: as a client side developer, you'd actually have to dive into the documentation or notice it when debugging or logging to see that you're actually getting more properties than expected in some cases? I mean, that's not unlike how it works in ORM frameworks or so, in my experience.

> P.S: Quite interestingly, but also obviously, when using type cast segments, @odata.type won't be present: [see this example](https://services.odata.org/V4/(S(xjqbds2oavibr01gt1fny24s))/TripPinServiceRW/People('russellwhyte')/Trips(1003)/PlanItems/Microsoft.OData.SampleService.Models.TripPin.Flight)

That is indeed quite interesting. It seems like @odata.type is only present when there's a possibility of the response containing multiple return types. When you require a certain sub type is used, then no other possibility is there.

If you have a concrete experiment in mind for how our (.Net) OData Service works in some edge cases, let me know. I'll see what I can do.

**Comment by texttechne** [comment]
Hi @brease-colin,

yeah was a bit lost there and wanted a sparrings partner, I think...

In any case, you should now get first class support for handling derived entity types.
Released with [@odata2ts/odata2ts@0.39.0](https://github.com/odata2ts/odata2ts/releases/tag/%40odata2ts%2Fodata2ts%400.39.0)

### Broader Topic of Handling Derived Entity Types
These issues are all connected to the broader topic of derived entity type support:
* #330 
* #333 
* #329: This allows you to dispense with the workaround regarding the expand
* this issue here, then is all about using the dedicated q-object for conversion by evaluating the `@odata.type` attribute

### Automatic Conversion by `@odata.type`
Each q-object representing a BaseType is now able to convert to its subtypes. As long as the correct `@odata.type` value is provided, then this conversion from and to the odata model works out-of-the-box.

This solves the given problen in this issue: Properties of subtypes will get converted!

### Feature Highlight
Especially the feature implemented in #330 is interesting. So for V4 services, every BaseType (meaning any ComplexType or EntityType which is referenced by any other type via the `BaseType` property) will get an additional casting method in its `EntityTypeService` and `EntitySetService`:
```
export class PlanItemCollectionService {
  // ...  
  public asFlightCollectionService() {
    const { client, path, options } = this.__base;
    return new FlightCollectionService(client, path, "Microsoft.OData.SampleService.Models.TripPin.Flight", {
      ...options,
      subtype: true,
    });
  }
  // ... 
}
```

This enables you to query, create, patch and update derived entity types. 
Benefit: The typing, especially the typing of response objects, will necessarily be correct.
```
// Retrieve only plan items which are flights => response: Array<FlightModel>
planItemCollectionService.asFlightCollectionService().query() 
// GET PlanItems/Microsoft.OData.SampleService.Models.TripPin.Flight

// create, patch and update will automatically add the correct @odata.type and not use the cast path segment
planItemCollectionService.asFlightCollectionService().create({ planItemId: 123, flightNumber: "F101" }) 
// POST PlanItems  
// { "@odata.type": "#Microsoft.OData.SampleService.Models.TripPin.Flight", PlanItemId: 123, FlightNumber: "F101" }

// but you could as well only use the cast past segment instead of the control info
planItemCollectionService.asFlightCollectionService().create({ planItemId: 123, flightNumber: "F101" }, undefined, { withCastPathSegment: true });
// POST PlanItems/Microsoft.OData.SampleService.Models.TripPin.Flight
// { PlanItemId: 123, FlightNumber: "F101" } 
```

So odata2ts will support you here, if you manage to provide the casting in this imperative fashion.
If you can't or won't, the you can still use the `PlanItemService`....

### Note on Cast Path Segments
It definitely depends on the backend implementation of the OData service whether cast path segments are supported and to what extent: See
[this SO answer](https://stackoverflow.com/questions/34278774/how-to-post-a-derived-entity-to-a-base-entity-set#answer-34280025)
